lib/solver.ml

```diff
@@ -98,20 +109,59 @@ let rec solve1 sol = function
   | Conj cs ->
      solve_conj sol cs
   | Disj cs ->
-     let issue_links = ["https://github.com/dwango/fialyzer/issues/99"] in
-     let message = "Solve disjunction of constraints" in
-     Error Known_error.(FialyzerError (NotImplemented {issue_links; message}))
+     solve_disj sol cs
 and solve_conj sol = function
   | [] -> Ok sol
   | c :: cs ->
      let open Result in
      solve1 sol c >>= fun sol' ->
      solve_conj sol' cs
+and solve_disj sol cs =
+  match List.filter_map ~f:(solve1 sol >>> Result.ok) cs with
+  | [] ->
+     let err0 = solve1 sol (List.hd_exn cs) in
+     err0
+  | sols ->
+     Ok (List.fold_left ~f:merge_solutions ~init:sol sols)
+
+(*
+  After calculating success typing, it is finally checked whether all branches pass type checking.
+  Note that this process is outside of the success typing algorithm.
+ *)
+let rec find_error_clauses sol = function
+  | Empty -> []
+  | Eq (ty1,ty2) ->
+     find_error_clauses sol (Subtype (ty1, ty2))
+     @ find_error_clauses sol (Subtype (ty2, ty1))
+  | Subtype (ty1, ty2) ->
+     begin match solve_sub sol ty1 ty2 with
+     | Ok _ -> []
+     | Error e -> [e]
+     end
+  | Disj cs ->
+     List.map ~f:(find_error_clauses sol) cs
+     |> List.concat
+  | Conj cs ->
+     find_conj sol cs
+and find_conj sol cs =
+  let f sol' c =
+    match (find_error_clauses sol' c, solve1 sol' c) with
```

Comment:
Would you tell me the reason why it iterates `find_error_clauses` with solution updating?
I think it is unnecessary because given solution is already updated by iteration of `solve1`.

Reply:
There is a clear reason for this. This is to prevent correct code from type errors.
For example,
```erlang
case N of
  0 -> N + 1;
  'foo' -> list_of_atom(N)
end.
```
is valid erlang code. The solution of the constraints for this code is:
```
...
τ_N ↦ 0 | 'foo'
...
```
If we do not update, it will type error in the clause `N + 1` since τ_N is not subtype of `number()`.

Reply:
I see! Thank you for your explanation in detail.